cheatsheets/OS_Command_Injection_Defense_Cheat_Sheet.md

```diff
@@ -58,11 +84,12 @@ If calling a system command that incorporates user-supplied cannot be avoided, t
 - In regards to the **arguments** used for these commands, they should be validated using the following options:
     - **Positive or allow list input validation**: Where are the arguments allowed explicitly defined.
     - **Allow list Regular Expression**: Where a list of good, allowed characters and the maximum length of the string are defined. Ensure that metacharacters like ones specified in `Note A` and white-spaces are not part of the Regular Expression. For example, the following regular expression only allows lowercase letters and numbers and does not contain metacharacters. The length is also being limited to 3-10 characters: `^[a-z0-9]{3,10}$`
+- According to **Guidline 10** of this [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html), *The first -- argument that is not an option-argument should be accepted as a delimiter indicating the end of options. Any following arguments should be treated as operands, even if they begin with the '-' character.* For example, `curl -- $url` will prevent an argument injection even if the `$url` is malformed and contains an additional argument.
 
 **Note A:**
 
 ```text
-& |  ; $ > < ` \ !
+& |  ; $ > < ` \ ! -
```

Comment:
Somewhere, we should note that the set of shell meta characters will depend on the specific shell that is being used. It should be obvious, but there are many who've only ever used a single command line shell.

Reply:
Yeah. like for example in `fish` there's no need for a `$(command)` just `(command)` work as command substitution. And `(` `)` these characters are not added in the list.

Reply:
+1

Reply:
@h4r1337 -  A few things things... 
1. If we wish to ensure that  the characters are truly excluded from any regex allow list as this earlier statement implies:
> Ensure that metacharacters like ones specified in `Note A` and white-spaces are not part of the Regular Expression.

Then  we should be tell them that they need to be careful with ranges in character classes that might accidentally include ranges. since the grave accent appears between uppercase and lowercase alphabetic characters in ASCII so some regular expression libraries could interpret something like '[A-z0-9]' as including the grave accent.

2. Why are we including '-' here at all? I'm not getting that one. I can introduce command flags, but that doesn't seem to be sufficient reason to me so I must be missing something. In ESAPI's `Encoder.encodeForOS()`, we consider '-' "immune from encoding" just as we do for alphanumeric characters.
3. I do think both `(` and `)` should be added to the list, not just because of fish, but for other shells such as bash, unmatched unquoted parens can be used as to cause syntax errors which could have unintended consequences or even in some cases be used as a DoS-like attack.
4. Whatever character is used to quote shell-metacharacters (typically '\') should be considered here as well.

Reply:
@kwwall I think adding `-` was a mistake from my side. Since we are adding `--` there's no need to include `-` in the list and this decision to add `-` in the list may not be appropriate in all contexts. So I will be removing it from the list.

Also adding single and double quotes and parantheses to the list as well.